add endpoint to stop jobs

shaneboulden opened the pull request:
This PR adds an endpoint to stop a test suite, including cleaning up Kubernetes batch job objects and all associated running pods. 

A few notes:

- There's presently no check to see if the suite is already complete. All test suites can be stopped - for those already complete, the Kubernetes batch job will simply deleted. For running jobs, the running pods will also be cleaned up.
- I haven't done any extensive testing on propagation policies. I'm using a 'Foreground' policy for now.
- There's still a 5-10 second delay between cancelling a job via the `peakorc` API, and the pods being cleaned up. I'm not sure if a different propagation policy will assist this.

shaneboulden committed:
add endpoint to stop jobs

adamgoossens commented:
Looks good to me. I don't know if the propagation policy will change the delay. We might be able to set a background policy for the Job, then delete the pods directly with additional API calls?

shaneboulden committed:
use background policy for deleting jobs

The `Foreground` propagation policy blocks the delete call until all
resources are deleted. By using a `Background` policy the job is
immediately deleted, and if we like we can clean up residual pods
manually.

shaneboulden commented:
There's a bit more clarity on the propagation policies [here](https://github.com/kubernetes/kubernetes/issues/20902).

It looks like the `Background` deletes the job upfront, and then cleans up the resources in the background. Whereas the `Foreground` policy blocks the delete call until the resources (pods) are cleaned up. 

It's possible the Foreground policy could end up in a weird state if we try directly cleaning up pods from under it. I think your way sounds best - have the `Background` policy initially clean up the job, and then make individual API calls to clean up the pods faster. I've just made the changes for the `Background` policy, I'll create an additional Trello ticket to clean up pods with individual API calls.

shaneboulden commented:
New trello card created here: https://trello.com/c/CBQpaOEA
